test/format/trace.c

```diff
@@ -0,0 +1,132 @@
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "format.h"
+
+#define TRACE_DIR "OPS.TRACE"
+#define TRACE_INIT_CMD "rm -rf %s/" TRACE_DIR " && mkdir %s/" TRACE_DIR
+
+int
+trace_config(const char *config)
+{
+    WT_DECL_RET;
+    char *copy, *p;
+
+    copy = dstrdup(config);
+    for (;;) {
+        if ((p = strstr(copy, "all")) != NULL) {
+            g.trace_all = true;
+            memset(p, ' ', strlen("all"));
+            continue;
+        }
+        if ((p = strstr(copy, "local")) != NULL) {
+            g.trace_local = true;
+            memset(p, ' ', strlen("local"));
+            continue;
+        }
+        break;
+    }
+
+    for (p = copy; *p != '\0'; ++p)
+        if (*p != ',' && !__wt_isspace((u_char)*p)) {
+            ret = EINVAL;
+            break;
+        }
+
+    free(copy);
+    return (ret);
+}
+
+void
+trace_init(void)
+{
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    size_t len;
+    char *p;
+    const char *config;
+
+    if (!g.trace)
+        return;
+
+    /* Write traces to a separate database by default, optionally write traces to the primary. */
+    if (g.trace_local) {
+        if (!g.c_logging)
+            testutil_die(EINVAL,
+              "operation logging to the primary database requires logging be configured for that "
+              "database");
+
+        conn = g.wts_conn;
+        testutil_check(conn->reconfigure(conn, "debug_mode=(log_retain=10)"));
+    } else {
+        len = strlen(g.home) * 2 + strlen(TRACE_INIT_CMD) + 10;
+        p = dmalloc(len);
+        testutil_check(__wt_snprintf(p, len, TRACE_INIT_CMD, g.home, g.home));
+        testutil_checkfmt(system(p), "%s", "logging directory creation failed");
+        free(p);
+
+        /* Configure log archival, and keep the last 20 log files. */
+        len = strlen(g.home) * strlen(TRACE_DIR) + 10;
+        p = dmalloc(len);
+        testutil_check(__wt_snprintf(p, len, "%s/%s", g.home, TRACE_DIR));
+        config = "create,log=(enabled,archive),debug_mode=(log_retain=10)";
+        testutil_checkfmt(wiredtiger_open(p, NULL, config, &conn), "%s: %s", p, config);
+        free(p);
+    }
+
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    g.trace_conn = conn;
+    g.trace_session = session;
```

Comment:
Ah, I understand now. That is why you don't want LSNs or they wouldn't be useful.

Reply:
Not sure I'm following, but if you're good, I'm good.

Reply:
I was thinking that LSNs would be useful so that you could look at specific operations around that LSN but since this is a separate connection, there is no way to connect something in the "real" connection/log with anything in this trace log. At the time I wrote that comment, I thought it was all intermingled in the log. That still might be an okay idea. I don't think that writing messages slows down logging or the test much.

Reply:
That is my intent: there's a flag in this change, `-L local` that puts format traces into the primary database log instead of a separate database. Once we do WT-6304 (put verbose messages into the log), then there's a single set of logs that contains verbose messages, format messages, and normal log records, so looking at the activity around a particular LSN (or timestamp), makes a lot of sense.

We could also hook it up so verbose messages go into the trace database I'm creating for format, but I think that's generally less interesting, I agree with you that adding messages to the log won't matter much.